Remove `:type` lines now sphinx-autoapi supports typehints

ashb opened the pull request:
Since we have no updated sphinx-autoapi to a more recent version it supports showing type hints in the documentation, so we don't need to have the type hints _and_ the `:type` lines -- which is good, as the ones in the doc strings are easy to get out of date!

The following settings have been set:

- `autodoc_typehints = 'description'` -- show types in description (where previous `:type` used to show up)

- `autodoc_typehints_description_target = 'documented'` -- only link to types that are documented. (Without this we have some missing return types that aren't documented, and aren't linked to in our current python API docs, so this caused a build failure)

- `autodoc_typehints_format = 'short'` -- Shorten type hints where possible, i.e. `StringIO` instead of `io.StringIO`

Before (https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/baseoperator/index.html#airflow.models.baseoperator.BaseOperator)

![image](https://user-images.githubusercontent.com/34150/150141676-6c6c5ab0-d595-4018-87a5-61794bd2dcab.png)


After

![image](https://user-images.githubusercontent.com/34150/150141626-ea21a80d-b89d-42b0-bc24-f687f704d193.png)


<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ashb committed:
Remove `:type` lines now sphinx-autoapi supports typehints

Since we have no updated sphinx-autoapi to a more recent version it
supports showing type hints in the documentation, so we don't need to
have the type hints _and_ the `:type` lines -- which is good, as the
ones in the doc strings are easy to get out of date!

The following settings have been set:

`autodoc_typehints = 'description'` -- show types in description (where
previous `:type` used to show up)

`autodoc_typehints_description_target = 'documented'` -- only link to
types that are documented. (Without this we have some missing return
types that aren't documented, and aren't linked to in our current python
API docs, so this caused a build failure)

`autodoc_typehints_format = 'short'` -- Shorten type hints where
possible, i.e. `StringIO` instead of `io.StringIO`

ashb commented on setup.cfg:
```diff
@@ -92,12 +92,6 @@ install_requires =
-    #      https://github.com/readthedocs/sphinx_rtd_theme/issues/1112
-    # This limitation can be removed after this issue is closed:
-    #      https://github.com/readthedocs/sphinx_rtd_theme/issues/1115
-    docutils<0.17
```
I don't think this one caused the backtrack, but it's not needed anymore either way.

potiuk replied:
I think the "lazy-object-proxy' in Dockerfile.ci might be the reason  - see #20979

ashb replied:
I really wish this process was more predictable @uranusjr (but I can't think of any obvious way of making it more so)

potiuk replied:
Agree. This ain't easy thing to do. But possibly making eager-update take into account the extras in "one pass" might be possible @uranusjr ? Not sure if that's possible, or if there are some reasons for that. I could possibly open an issue for that, but I am afraid my issues might not be too welcomed as my persistency in improving warning message in the `root` case is seen as a bit `toxic` (even if it is a real issue to tackle).

ashb replied:
Anyway, I from localy testing it appears that the backtrack is removed just by adding the restriction on importlib-metadata, and isn't anything to do with removing this docutils

potiuk replied:
importlib-metadata was my next guess :)

ashb replied:
Either way, 3.7 image is pushing now so found and fixed the backtrack

ashb committed:
Add argument type names to local spelling dictionary

Now that we are using the type hints in the docs, sphinxcontrib-spelling
picks them up as words to be checked, so we have to ignore them.

I've chosen to add the provider specific ones to local dictionary files
rather than the global, as for example, `mgmt` is an error in most
places, but not in some of the Azure provider.
